src/librustc_lint/unused.rs

```diff
@@ -464,3 +466,64 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedAllocation {
         }
     }
 }
+
+declare_lint! {
+    pub(super) UNUSED_LOOP_LABEL,
+    Warn,
+    "warns on unused labels for loops"
+}
+
+#[derive(Clone)]
+pub struct UnusedLoopLabel(pub vec::Vec<ast::Label>);
+
+impl UnusedLoopLabel {
+    pub fn new() -> Self {
+        UnusedLoopLabel(vec![])
+    }
+}
+
+impl LintPass for UnusedLoopLabel {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_LOOP_LABEL)
+    }
+}
+
+impl EarlyLintPass for UnusedLoopLabel {
+    fn check_expr(&mut self, _: &EarlyContext, expr: &ast::Expr) {
+        match expr.node {
+            ast::ExprKind::While(_, _, Some(ref label))
+            | ast::ExprKind::WhileLet(_, _, _, Some(ref label))
+            | ast::ExprKind::ForLoop(_, _, _, Some(ref label))
+            | ast::ExprKind::Loop(_, Some(ref label)) => {
+                self.0.push(*label);
+            }
+            ast::ExprKind::Break(Some(ref label), _) | ast::ExprKind::Continue(Some(ref label)) => {
+                'remove_used_label: for i in (0..self.0.len()).rev() {
+                    if self.0.get(i).unwrap().ident.name == label.ident.name {
```

Comment:
Comparing `name`s is not correct in general because macro hygiene applies to labels as well.
Еhis lint should not repeat the name resolution pass (`librustc_resolve/lib.rs`), it should become a part of it - each label definition (`fn with_resolved_label`) should add that label's `id` into "unused" set and then each label use (`fn search_label`) should remove the definition it resolves to from the set.